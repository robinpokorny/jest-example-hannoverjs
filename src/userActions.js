export const LOAD_USER_BOOKINGS = 'LOAD_USER_BOOKINGS'

// Expects that `fetch` is injected using `extraArgument`
// while creating the store:
//   const store = createStore(
//     reducer,
//     applyMiddleware(thunk.withExtraArgument({ fetch }))
//   )
// Read more:
// https://github.com/gaearon/redux-thunk#injecting-a-custom-argument
//
// The created action will be evaluated by redux-promise-middleware:
// https://github.com/pburtchaell/redux-promise-middleware

export const loadUserBookings = () =>
  (dispatch, getState, { fetch }) => {
    const { name } = getState()

    const promise = fetch(`${name}/bookings`)
      .then((response) =>
        response.json()
      )

    dispatch({
      type: LOAD_USER_BOOKINGS,
      payload: {
        data: name,
        promise
      }
    })
  }
