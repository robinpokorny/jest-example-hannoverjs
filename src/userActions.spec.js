import { isFSA } from 'flux-standard-action'

import * as actions from './userActions'

describe('loadUserBookings', () => {
  const action = actions.loadUserBookings()

  const dispatch = jest.fn()
  const getState = () => ({
    name: 'gregor'
  })
  const deps = {
    fetch: jest.fn(() => Promise.resolve({
      json: () => {}
    }))
  }

  beforeEach(() => {
    dispatch.mockClear()
    deps.fetch.mockClear()
  })

  test('is a valid FSA', () => {
    action(dispatch, getState, deps)

    expect(dispatch).toHaveBeenCalledTimes(1)
    const dispatchedAction = dispatch.mock.calls[0][0]

    expect(isFSA(dispatchedAction)).toBe(true)
  })

  test('has correct data', () => {
    action(dispatch, getState, deps)

    expect(dispatch.mock.calls).toMatchSnapshot()
  })

  test('fetches bookings from the server', () => {
    action(dispatch, getState, deps)

    expect(deps.fetch).toHaveBeenCalledTimes(1)
    expect(deps.fetch.mock.calls).toMatchSnapshot()
  })
})
