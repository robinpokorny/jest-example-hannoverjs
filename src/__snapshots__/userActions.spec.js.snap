// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadUserBookings fetches bookings from the server 1`] = `
Array [
  Array [
    "gregor/bookings",
  ],
]
`;

exports[`loadUserBookings has correct data 1`] = `
Object {
  "payload": Object {
    "data": "gregor",
    "promise": Promise {},
  },
  "type": "LOAD_USER_BOOKINGS",
}
`;
